File: src/spec/resources/testprojects/javatest/src/main/java/com/example/petstore/Configuration.java
package com.example.petstore;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import javax.annotation.Nullable;

/**
 * Configuration for API clients.
 *
 * <p>This class is immutable. Use {@link #builder()} to create instances:
 *
 * <pre>{@code
 * Configuration config = Configuration.builder()
 *     .baseUrl("https://api.example.com")
 *     .defaultHeader("Authorization", "Bearer token")
 *     .verifySsl(false)
 *     .build();
 * }</pre>
 */
public final class Configuration {

  @Nullable private static volatile Configuration defaultInstance;

  private final String baseUrl;
  private final Map<String, String> defaultHeaders;
  private final boolean debug;
  private final boolean verifySsl;
  @Nullable private final String sslCaCert;
  @Nullable private final String certFile;
  @Nullable private final String keyFile;
  @Nullable private final String proxy;
  @Nullable private final Integer timeout;
  @Nullable private final Integer retries;

  Configuration(
      String baseUrl,
      Map<String, String> defaultHeaders,
      boolean debug,
      boolean verifySsl,
      @Nullable String sslCaCert,
      @Nullable String certFile,
      @Nullable String keyFile,
      @Nullable String proxy,
      @Nullable Integer timeout,
      @Nullable Integer retries) {
    this.baseUrl = baseUrl;
    this.defaultHeaders = Collections.unmodifiableMap(new HashMap<>(defaultHeaders));
    this.debug = debug;
    this.verifySsl = verifySsl;
    this.sslCaCert = sslCaCert;
    this.certFile = certFile;
    this.keyFile = keyFile;
    this.proxy = proxy;
    this.timeout = timeout;
    this.retries = retries;
  }

  /**
   * Create a new builder for constructing Configuration instances.
   *
   * @return a new builder
   */
  public static Builder builder() {
    return new Builder();
  }

  /**
   * Return the default configuration instance.
   *
   * @return the default configuration
   */
  public static Configuration getDefault() {
    if (defaultInstance == null) {
      defaultInstance = builder().build();
    }
    return defaultInstance;
  }

  /**
   * Set the default configuration instance.
   *
   * @param config the configuration to use as default
   */
  public static void setDefault(Configuration config) {
    defaultInstance = config;
  }

  public String getBaseUrl() {
    return baseUrl;
  }

  public Map<String, String> getDefaultHeaders() {
    return defaultHeaders;
  }

  public boolean isDebug() {
    return debug;
  }

  public boolean isVerifySsl() {
    return verifySsl;
  }

  @Nullable
  public String getSslCaCert() {
    return sslCaCert;
  }

  @Nullable
  public String getCertFile() {
    return certFile;
  }

  @Nullable
  public String getKeyFile() {
    return keyFile;
  }

  @Nullable
  public String getProxy() {
    return proxy;
  }

  @Nullable
  public Integer getTimeout() {
    return timeout;
  }

  @Nullable
  public Integer getRetries() {
    return retries;
  }

  /** Builder for creating immutable {@link Configuration} instances. */
  public static final class Builder {

    private String baseUrl = "/api/v3";
    private final Map<String, String> defaultHeaders = new HashMap<>();
    private boolean debug = false;
    private boolean verifySsl = true;
    @Nullable private String sslCaCert = null;
    @Nullable private String certFile = null;
    @Nullable private String keyFile = null;
    @Nullable private String proxy = null;
    @Nullable private Integer timeout = null;
    @Nullable private Integer retries = null;

    Builder() {}

    /** Set the base URL for all API requests. */
    public Builder baseUrl(String baseUrl) {
      this.baseUrl = baseUrl;
      return this;
    }

    /** Add a default header to include in every API request. */
    public Builder defaultHeader(String name, String value) {
      this.defaultHeaders.put(name, value);
      return this;
    }

    /** Set all default headers to include in every API request. */
    public Builder defaultHeaders(Map<String, String> headers) {
      this.defaultHeaders.putAll(headers);
      return this;
    }

    /** Enable or disable debug logging. */
    public Builder debug(boolean debug) {
      this.debug = debug;
      return this;
    }

    /** Enable or disable SSL/TLS certificate verification. */
    public Builder verifySsl(boolean verifySsl) {
      this.verifySsl = verifySsl;
      return this;
    }

    /** Set the path to a CA certificate file for SSL/TLS verification. */
    public Builder sslCaCert(@Nullable String sslCaCert) {
      this.sslCaCert = sslCaCert;
      return this;
    }

    /** Set the path to a client certificate file for mutual TLS. */
    public Builder certFile(@Nullable String certFile) {
      this.certFile = certFile;
      return this;
    }

    /** Set the path to a client private key file for mutual TLS. */
    public Builder keyFile(@Nullable String keyFile) {
      this.keyFile = keyFile;
      return this;
    }

    /** Set the proxy URL for all API requests. */
    public Builder proxy(@Nullable String proxy) {
      this.proxy = proxy;
      return this;
    }

    /** Set the request timeout in seconds. */
    public Builder timeout(@Nullable Integer timeout) {
      this.timeout = timeout;
      return this;
    }

    /** Set the number of retry attempts for failed requests. */
    public Builder retries(@Nullable Integer retries) {
      this.retries = retries;
      return this;
    }

    /** Build and return an immutable Configuration instance. */
    public Configuration build() {
      return new Configuration(
          baseUrl,
          new HashMap<>(defaultHeaders),
          debug,
          verifySsl,
          sslCaCert,
          certFile,
          keyFile,
          proxy,
          timeout,
          retries);
    }
  }
}

File: src/spec/resources/testprojects/pytest/petstore_client/configuration.py
from dataclasses import dataclass, field
from types import MappingProxyType
from typing import ClassVar, Dict, Mapping, Optional

from typing_extensions import Self


@dataclass(frozen=True)
class Configuration:
    """Configuration for API clients.

    Holds settings that apply to all API requests such as the base URL,
    default headers, TLS options, proxy, timeout, and retry policy.

    This class is immutable. Use :meth:`builder` to create instances::

        config = Configuration.builder() \\
            .base_url('https://api.example.com') \\
            .default_header('Authorization', 'Bearer token') \\
            .verify_ssl(False) \\
            .build()
    """

    base_url: str = '/api/v3'
    default_headers: Mapping[str, str] = field(default_factory=lambda: MappingProxyType({}))
    debug: bool = False
    verify_ssl: bool = True
    ssl_ca_cert: Optional[str] = None
    cert_file: Optional[str] = None
    key_file: Optional[str] = None
    proxy: Optional[str] = None
    timeout: Optional[int] = None
    retries: Optional[int] = None

    _default: ClassVar[Optional['Configuration']] = None

    def __post_init__(self) -> None:
        if not isinstance(self.default_headers, MappingProxyType):
            object.__setattr__(self, 'default_headers', MappingProxyType(dict(self.default_headers)))

    @classmethod
    def get_default(cls) -> 'Configuration':
        """Return the default configuration instance, creating it lazily if needed."""
        if cls._default is None:
            cls._default = cls()
        return cls._default

    @classmethod
    def set_default(cls, configuration: Optional[Self]) -> None:
        """Set the default configuration instance."""
        cls._default = configuration

    @classmethod
    def builder(cls) -> 'ConfigurationBuilder':
        """Create a new builder for constructing Configuration instances."""
        return ConfigurationBuilder()


class ConfigurationBuilder:
    """Builder for creating immutable :class:`Configuration` instances.

    Example::

        config = Configuration.builder() \\
            .base_url('https://api.example.com') \\
            .verify_ssl(False) \\
            .build()
    """

    def __init__(self) -> None:
        self._base_url: str = '/api/v3'
        self._default_headers: Dict[str, str] = {}
        self._debug: bool = False
        self._verify_ssl: bool = True
        self._ssl_ca_cert: Optional[str] = None
        self._cert_file: Optional[str] = None
        self._key_file: Optional[str] = None
        self._proxy: Optional[str] = None
        self._timeout: Optional[int] = None
        self._retries: Optional[int] = None

    def base_url(self, base_url: str) -> 'ConfigurationBuilder':
        """Set the base URL for all API requests."""
        self._base_url = base_url
        return self

    def default_header(self, name: str, value: str) -> 'ConfigurationBuilder':
        """Add a default header to include in every API request."""
        self._default_headers[name] = value
        return self

    def default_headers(self, headers: Dict[str, str]) -> 'ConfigurationBuilder':
        """Set all default headers to include in every API request."""
        self._default_headers.update(headers)
        return self

    def debug(self, debug: bool) -> 'ConfigurationBuilder':
        """Enable or disable debug logging."""
        self._debug = debug
        return self

    def verify_ssl(self, verify_ssl: bool) -> 'ConfigurationBuilder':
        """Enable or disable SSL/TLS certificate verification."""
        self._verify_ssl = verify_ssl
        return self

    def ssl_ca_cert(self, ssl_ca_cert: Optional[str]) -> 'ConfigurationBuilder':
        """Set the path to a CA certificate file for SSL/TLS verification."""
        self._ssl_ca_cert = ssl_ca_cert
        return self

    def cert_file(self, cert_file: Optional[str]) -> 'ConfigurationBuilder':
        """Set the path to a client certificate file for mutual TLS."""
        self._cert_file = cert_file
        return self

    def key_file(self, key_file: Optional[str]) -> 'ConfigurationBuilder':
        """Set the path to a client private key file for mutual TLS."""
        self._key_file = key_file
        return self

    def proxy(self, proxy: Optional[str]) -> 'ConfigurationBuilder':
        """Set the proxy URL for all API requests."""
        self._proxy = proxy
        return self

    def timeout(self, timeout: Optional[int]) -> 'ConfigurationBuilder':
        """Set the request timeout in seconds."""
        self._timeout = timeout
        return self

    def retries(self, retries: Optional[int]) -> 'ConfigurationBuilder':
        """Set the number of retry attempts for failed requests."""
        self._retries = retries
        return self

    def build(self) -> Configuration:
        """Build and return an immutable Configuration instance."""
        return Configuration(
            base_url=self._base_url,
            default_headers=self._default_headers,
            debug=self._debug,
            verify_ssl=self._verify_ssl,
            ssl_ca_cert=self._ssl_ca_cert,
            cert_file=self._cert_file,
            key_file=self._key_file,
            proxy=self._proxy,
            timeout=self._timeout,
            retries=self._retries,
        )
